src/transformers/models/whisper/modeling_whisper.py

```diff
@@ -2555,22 +2556,71 @@ def _extract_token_timestamps(self, generate_outputs, alignment_heads, time_prec
+            beam_indices = generate_outputs.beam_indices[:, :weight_length]
+            beam_indices = beam_indices.masked_fill(beam_indices == -1, 0)
+
+            # Select the cross attention from the right beam for each output sequences
+            weights = torch.stack(
+                [
+                    torch.index_select(weights[:, :, i, :], dim=0, index=beam_indices[:, i])
+                    for i in range(beam_indices.shape[1])
+                ],
+                dim=2,
+            )
+
+        timestamps = torch.zeros_like(generate_outputs.sequences, dtype=torch.float32)
+        batch_size = timestamps.shape[0]
+
         if num_frames is not None:
-            weights = weights[..., : num_frames // 2]
+            # two cases:
+            # 1. num_frames is the same for each sample -> compute the DTW matrix for each sample in parallel
+            # 2. num_frames is different, compute the DTW matrix for each sample sequentially
+
+            # we're using np.unique because num_frames can be int/list/tuple
+            if len(np.unique(num_frames)) == 1:
+                # if num_frames is the same, no need to recompute matrix, std and mean for each element of the batch
+                num_frames = num_frames if isinstance(num_frames, int) else num_frames[0]
 
-        # Normalize and smoothen the weights.
-        std, mean = torch.std_mean(weights, dim=-2, keepdim=True, unbiased=False)
-        weights = (weights - mean) / std
-        weights = _median_filter(weights, self.config.median_filter_width)
+                weights = weights[..., : num_frames // 2]
+            else:
+                # num_frames is of shape (batch_size,) whereas batch_size is truely batch_size*num_return_sequences
+                repeat_time = batch_size if isinstance(num_frames, int) else batch_size // len(num_frames)
+                num_frames = np.repeat(num_frames, repeat_time)
 
-        # Average the different cross-attention heads.
-        matrix = weights.mean(dim=1)
+        if num_frames is None or isinstance(num_frames, int):
+            # Normalize and smoothen the weights.
+            std, mean = torch.std_mean(weights, dim=-2, keepdim=True, unbiased=False)
+            weights = (weights - mean) / std
+            weights = _median_filter(weights, self.config.median_filter_width)
 
-        timestamps = torch.zeros_like(generate_outputs.sequences, dtype=torch.float32)
+            # Average the different cross-attention heads.
+            weights = weights.mean(dim=1)
 
         # Perform dynamic time warping on each element of the batch.
-        for batch_idx in range(timestamps.shape[0]):
-            text_indices, time_indices = _dynamic_time_warping(-matrix[batch_idx].double().cpu().numpy())
+        for batch_idx in range(batch_size):
+            if num_frames is not None and isinstance(num_frames, (tuple, list)):
+                matrix = weights[batch_idx, ..., : num_frames[batch_idx] // 2]
```

Comment:
Good for me smoothing the cross-attention weights matrix sequentially (rather than batched) since we have to do it this way for DTW anyway